rinja_parser/src/lib.rs

```diff
@@ -962,35 +947,75 @@ impl<'a> SyntaxBuilder<'a> {
     }
 }
 
-#[derive(Clone, Copy, Default)]
-pub(crate) struct Level(u8);
+#[derive(Debug, Clone, Copy)]
+struct Level<'l>(&'l Cell<usize>);
+
+impl Level<'_> {
+    const MAX_DEPTH: usize = 128;
 
-impl Level {
-    fn nest(self, i: &str) -> ParseResult<'_, Level> {
-        if self.0 >= Self::MAX_DEPTH {
-            return Err(winnow::error::ErrMode::Cut(ErrorContext::new(
-                "your template code is too deeply nested, or last expression is too complex",
-                i,
-            )));
+    fn guard(&self) -> LevelGuard<'_> {
+        LevelGuard {
+            level: *self,
+            count: 0,
         }
+    }
 
-        Ok(Level(self.0 + 1))
+    fn nest<'a>(&self, i: &'a str) -> ParseResult<'a, LevelGuard<'_>> {
+        if let Some(new_level) = self.0.get().checked_sub(1) {
+            self.0.set(new_level);
+            Ok(LevelGuard {
+                level: *self,
+                count: 1,
+            })
+        } else {
+            Err(Self::_fail(i))
+        }
     }
 
-    const MAX_DEPTH: u8 = 128;
+    #[cold]
+    #[inline(never)]
+    fn _fail(i: &str) -> ParseErr<'_> {
+        winnow::error::ErrMode::Cut(ErrorContext::new(
+            "your template code is too deeply nested, or the last expression is too complex",
+            i,
+        ))
+    }
+}
+
+#[must_use]
+struct LevelGuard<'l> {
```

Comment:
Please add some code explanations about why we have this type (like what you wrote in the PR description). Future us will thank us. :wink:

Reply:
Good call. :) Added.